crates/rover-client/src/operations/subgraph/fetch_all/runner.rs

```diff
@@ -37,76 +38,162 @@ fn get_subgraphs_from_response_data(
-                    can_operation_convert: true,
-                })
-            },
-            |subgraphs| {
-                Ok(subgraphs
-                    .into_iter()
-                    .map(|subgraph| {
-                        Subgraph::builder()
-                            .name(subgraph.name.clone())
-                            .and_url(subgraph.url)
-                            .sdl(subgraph.active_partial_schema.sdl)
-                            .build()
-                    })
-                    .collect())
-            },
-        ),
+        Some(SubgraphFetchAllGraphVariant::GraphVariant(variant)) => {
+            extract_subgraphs_from_response(variant, input.graph_ref)
+        }
         _ => Err(RoverClientError::InvalidGraphRef),
     }
 }
+fn extract_subgraphs_from_response(
+    value: SubgraphFetchAllQueryVariantOnGraphVariant,
+    graph_ref: GraphRef,
+) -> Result<Vec<Subgraph>, RoverClientError> {
+    match (value.subgraphs, value.source_variant) {
+        // If we get null back in both branches or the query, or we get a structure in the
+        // sourceVariant half but there are no subgraphs in it. Then we return an error
+        // because this isn't a FederatedSubgraph **as far as we can tell**.
+        (None, None)
+        | (
+            None,
+            Some(SubgraphFetchAllQueryVariantOnGraphVariantSourceVariant { subgraphs: None }),
+        ) => Err(RoverClientError::ExpectedFederatedGraph {
+            graph_ref,
+            can_operation_convert: true,
+        }),
+        // If we get nothing from querying the subgraphs directly, but we do get some subgraphs
+        // on the sourceVariant side of the query, we just return those.
+        (
+            None,
+            Some(SubgraphFetchAllQueryVariantOnGraphVariantSourceVariant {
+                subgraphs: Some(subgraphs),
+            }),
+        ) => Ok(subgraphs
+            .into_iter()
+            .map(|subgraph| subgraph.into())
+            .collect()),
+        // Here there are three cases where we might want to return the subgraphs we got from
+        // directly querying the graphVariant:
+        // 1. If we get subgraphs back from the graphVariant directly and nothing from the sourceVariant
+        // 2. If we get subgraphs back from the graphVariant directly and a structure from the
+        // sourceVariant, but it contains no subgraphs
+        // 3. If we get subgraphs back from both 'sides' of the query, we take the results from
+        // querying the **graphVariant**, as this is closest to the original behaviour, before
+        // we introduced the querying of the sourceVariant.
+        (Some(subgraphs), None)
+        | (
+            Some(subgraphs),
+            Some(SubgraphFetchAllQueryVariantOnGraphVariantSourceVariant { subgraphs: None }),
+        )
+        | (
+            Some(subgraphs),
+            Some(SubgraphFetchAllQueryVariantOnGraphVariantSourceVariant { subgraphs: Some(_) }),
```

Comment:
__NIT__: should we replace this with `(Some(subgraphs), _)`?
```suggestion
        (Some(subgraphs), _
```